include/isc_nav/utility/rrt.hpp

```diff
@@ -0,0 +1,267 @@
+                    }
+
+                    G.at(attach_from).at(attach_to)++;
+                    G.push_back(E);
+                }
+
+                // If vertex at goal return graph
+                if ((Xnew.x - goal_.x < 0.1) && (Xnew.y - goal_.y < 0.1))
+                {
+                    trace_back_path(Xnew, path);
+                }
+            }
+
+            return path;
+        }
+
+
+        Point2D randomPosition()
+        {
+            srand(time(0));
+
+            while (true) 
+            {
+                int32_t x = fRand(0, map_.get_size_x());
+                int32_t y = fRand(0, map_.get_size_y());
+
+                if (map_.valid_cell(x, y)) 
+                {
+                    Point2D random_node(x, y);
+                    return random_node;
+                }
+            }
+        }
+
+
+        Point2D findNearest(std::vector<Point2D>& V, Point2D random_node, double& Xnearest_dist)
+        {
+            double dist;
+            Point2D Xnearest;
+
+            if (!V.empty()) 
+            {
+                // Assign values to be compared
+                Xnearest_dist = getDistance(V.at(0), random_node);
+                Xnearest = V.at(0);
+            }
+
+            for (auto vertex : V)
+            {
+                dist = getDistance(vertex, random_node);
+                if (dist < Xnearest_dist)
+                {
+                    Xnearest_dist = dist;
+                    Xnearest = vertex;
+                }
+            }
+
+            return Xnearest;
+        }
+
+        // p1 is Xnearest, p2 is random node
+        Point2D findNew(Point2D& p1, Point2D& p2, double& Xnearest_dist)
+        {
+            Point2D Xnew;
+            double n = max_dist_;
+            double m = Xnearest_dist - n;
+
+            Xnew.x = (m * p2.x + n * p1.x)/(m + n);
+            Xnew.y = (m * p2.y + n * p1.y)/(m + n);
+
+            return Xnew;
+        }
+
+
+        bool isInObstacle(Point2D& random_node)
+        {
+            return map_.at(random_node) == CostMap::LETHAL_OBSTACLE;
+        }
+
+        bool isCollisionFree(Point2D& p1, Point2D& p2)
+        {
+            float reso = map_.get_resolution();
+            double dist = getDistance(p1, p2);
+            int num_of_pt = int(std::ceil(dist / reso));
+            Point2D pt = p1;
+
+            for(int i = 0; i < num_of_pt; i++)
+            {
+                pt = generateNew(pt, p2, dist, reso);
+                if (isInObstacle(pt))
+                {
+                    return false;
+                }
+            }
+            
+            return true;
+        }
+
+        Point2D generateNew(Point2D& p1, Point2D& p2, double& dist, float& reso)
+        {
+            Point2D Xnew;
+            double n = reso;
+            double m = dist - n;
+
+            Xnew.x = (m * p2.x + n * p1.x)/(m + n);
+            Xnew.y = (m * p2.y + n * p1.y)/(m + n);
+
+            dist -= reso;
```

Comment:
This updates dist externally because of the mutable reference. Not sure if this was intentional, so let this serve as a hint if you're getting weird behavior. 

This kind of thing shows up in other functions as well, so I would const all refrences (unless mutability was intentional) which will make updates like this a compiler error, pointing out any unintentional mutations.